plugin/entrySchema.go

```diff
@@ -259,11 +258,55 @@ func (s *EntrySchema) schemaOf(obj interface{}) (*JSONSchema, error) {
 // Helper for s.fill(). We make it a separate method to avoid re-creating
 // closures for each recursive s.fill() call.
 func (s *EntrySchema) fillPanicf(format string, a ...interface{}) {
-	formatStr := fmt.Sprintf("s.fill (%v): %v", s.TypeID, format)
+	formatStr := fmt.Sprintf("s.fill (%v): %v", TypeID(s.entry), format)
 	msg := fmt.Sprintf(formatStr, a...)
 	panic(msg)
 }
 
+func pluginName(e Entry) string {
+	// Using ID(e) will panic if e.id() is empty. The latter's possible
+	// via something like "Schema(registry) => TypeID(Root)", where
+	// CachedList(registry) was not yet called. This can happen if, for
+	// example, the user starts the Wash shell and runs `stree`.
+	trimmedID := strings.Trim(e.id(), "/")
+	if trimmedID == "" {
+		switch e.(type) {
+		case Root:
+			return CName(e)
+		case *Registry:
+			return ""
+		default:
+			// e has no ID. This is possible if e's from the apifs package. For now,
+			// it is enough to return "__apifs__" here because this is an unlikely
+			// edge case.
+			//
+			// TODO: Panic here once apifs is removed.
+			return "__apifs__"
+		}
+	}
+	segments := strings.Split(trimmedID, "/")
+	return segments[0]
+}
+
+func rawTypeID(e Entry) string {
+	switch t := e.(type) {
+	case externalPlugin:
+		rawTypeID := t.RawTypeID()
+		if rawTypeID == "" {
+			rawTypeID = "unknown"
+		}
+		return rawTypeID
+	default:
+		// e is either a core plugin entry or the plugin registry itself
+		reflectType := unravelPtr(reflect.TypeOf(e))
+		return reflectType.PkgPath() + "/" + reflectType.Name()
```

Comment:
Per https://github.com/puppetlabs/wash/pull/437#issuecomment-516696489, I'd like this to return `reflectType.Name()`. However with volume files/dirs, this would mean we'd get `dir`, `file` for the raw type IDs so the overall type ID would be `docker::dir`, `docker::file`, which isn't readable. Changing `volume.dir` => `volume.volumeDir`, `volume.file` => `volume.volumeFile` would be one way to solve it.

Reply:
How would using `reflectType.Name()` show up for other types? I'm not sure what the trade-offs are.

Reply:
I'm not sure where I'd see these naming differences in everyday use.

Reply:
Hmm, I was originally thinking of extracting the resource type from the type ID, but I think I'm leaning towards adding a core-plugin specific builder to EntrySchema that lets plugin authors set the resource type. That way, if a plugin has multiple packages for its resources (e.g. in AWS, we have an EC2 package for EC2 stuff; S3 package for S3 stuff), the analytics side of things will remain unchanged (resource type makes things stable against a massive code-organization-related refactor).

That would make most of this PR unnecessary (except maybe pluginName), although I do still need it to properly implement the kind primary so it may as well go in.